src/go/cmd/hw-exporter/main.go

```diff
@@ -0,0 +1,130 @@
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var (
+	metricsPort = flag.Int("metrics-port", 9999, "Port to expose Prometheus metrics on.")
+	logLevel    = flag.Int("log-level", int(slog.LevelInfo), "the log message level required to be logged")
+	chroot      = flag.String("chroot", "/", "Path to chroot into before collecting hardware info.")
+)
+
+type pciCollector struct {
+	pciDeviceCount *prometheus.Desc
+}
+
+func newPciCollector() *pciCollector {
+	return &pciCollector{
+		pciDeviceCount: prometheus.NewDesc(
+			"pci_device_count",
+			"Number of PCI devices by vendor, product, class, and driver.",
+			[]string{"vendor", "product", "class", "driver"},
+			nil,
+		),
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *pciCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.pciDeviceCount
+}
+
+// getNameOrID returns the name of a PCI device component, or its ID if the name is unknown.
+func getNameOrID(name, id string) string {
+	if name == util.UNKNOWN {
+		return "0x" + id
+	}
+	return name
+}
+
+// Collect implements the prometheus.Collector interface.
+func (c *pciCollector) Collect(ch chan<- prometheus.Metric) {
+	pciInfo, err := ghw.PCI(&option.Option{Chroot: chroot})
+	if err != nil {
+		slog.Error("Failed to get PCI info", ilog.Err(err))
+		return
+	}
+
+	deviceCounts := make(map[[4]string]float64)
+	for _, device := range pciInfo.Devices {
+		vendor := getNameOrID(device.Vendor.Name, device.Vendor.ID)
+		product := getNameOrID(device.Product.Name, device.Product.ID)
+		class := getNameOrID(device.Class.Name, device.Class.ID)
+		labels := [4]string{vendor, product, class, device.Driver}
+		deviceCounts[labels]++
+	}
+
+	for labels, count := range deviceCounts {
+		ch <- prometheus.MustNewConstMetric(c.pciDeviceCount, prometheus.GaugeValue, count, labels[0], labels[1], labels[2], labels[3])
+	}
+}
+
+func runMetricsServer(ctx context.Context) {
+
+}
+
+func main() {
+	flag.Parse()
+	logHandler := ilog.NewLogHandler(slog.Level(*logLevel), os.Stderr)
+	slog.SetDefault(slog.New(logHandler))
+
+	// Run once on startup to test container setup, this is useful during development.
+	_, err := ghw.PCI(&option.Option{Chroot: chroot})
+	if err != nil {
+		slog.Error("Failed to get PCI info", ilog.Err(err))
+		os.Exit(1)
+	}
+
+	// Handle SIGTERM for graceful shutdown.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		sigChan := make(chan os.Signal, 1)
+		signal.Notify(sigChan, syscall.SIGTERM)
+		<-sigChan
+		slog.Info("Received SIGTERM, shutting down...")
+		cancel()
+	}()
+
+	// Construct and run the metrics server.
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(newPciCollector())
+
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.HandlerFor(registry, promhttp.HandlerOpts{}))
+
+	server := &http.Server{
+		Addr:    fmt.Sprintf(":%d", *metricsPort),
+		Handler: mux,
+	}
+
+	go func() {
```

Comment:
I'd prefer swapping the goroutine and the main work here.

I.e. the main task of the process could (Should?) be in the main routine, while the goroutine would wait for the context to be Done() and handles the shutdown as pseudo-signal handler.